### ISSUE TITLE
read-char-no-hang slow on macOS

### ISSUE THREAD

**Issue by xrme** [Clozure/ccl#518]
Advisability of this usage aside, consider the following snippet:
```
(in-package :cl-user)

(defparameter *console-stream* nil)

(format t "~&Waiting for connection...")
(let* ((socket (ccl:make-socket :connect :passive ;; Listen
                                :reuse-address t
                                :nodelay t
                                :local-host "localhost"
                                :local-port 6444))
       
       (stream (ccl:accept-connection socket)))
  (setf *console-stream* stream))
(format t "connected!~%")

(time (loop for i from 1 to 10000000
            do
            (read-char-no-hang *console-stream*)))
```

Load this file and connect via `nc 127.0.0.1 6444` or some appropriate equivalent.  On FreeBSD, Linux, and OmniOS, this takes a handful of seconds to run after a client is connected.  A Mac takes so long that I got tired of waiting.

In Activity Monitor, the dx86cl64 process shows tens of thousands of "idle wake ups", which is abnormal, but it's not clear to me what idle wake ups are exactly, or what the large number of them signifies.

We use the poll(2) system call in ccl::fd-input-available-p, and this appears to be slow on the Mac for some reason.

https://gist.github.com/xrme/cd42d3d753ac00861e439e012366f2cf is a C program that demonstrates the difference in performance between poll(2) and select(2).  Compile and run and connect via `nc 127.0.0.1 6444` to see performance with poll.  Re-compile with `-DUSE_SELECT` and do the same to see performance with select.  On my machine, I get about 15 seconds with poll, and about 550ms with select.

This leads me to think that ccl::fd-input-available-p (and presumably ccl::fd-ready-for-output-p too) should use select instead of poll for `#+darwin-target`.

There will be a bootstrapping issue here: if we want to reference the `#>fd_set` type at this point (when l1-streams loads during bootstrapping, I mean), we'll need to update `*canonical-os-foreign-types*` in lib/foreign-types.lisp first.

In other words, it will be necessary to:
1. update `*canonical-os-foreign-types*` and rebuild-ccl.
2. now update fd-input-available-p to use `#>fd_set` and rebuild-ccl again.

Otherwise, the bootstrapping load will fail.

**Comment by xrme** [comment]
I posted https://developer.apple.com/forums/thread/764698 about this, and also sent a bug report to Apple about the performance difference.

Because this usage (calling poll with a zero timeout in a loop like this) is rather ill-considered, I prefer to stick to poll for simplicity & uniformity of implementation. We can encourage more suitable approaches (e.g., do a blocking read in another thread, and place results on a shared queue) for situations like this.
